edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnectionProvider.cs

```diff
@@ -177,6 +196,102 @@ public async Task<Try<ICloudConnection>> Connect(IIdentity identity, Action<stri
+        {
+            Events.CreatingCloudConnectionOnBehalfOf(identity);
+            ConnectionMetadata connectionMetadata = await this.metadataStore.GetMetadata(identity.Id);
+            string productInfo = connectionMetadata.EdgeProductInfo;
+            Option<string> modelId = connectionMetadata.ModelId;
+            try
+            {
+                ICloudConnection cc = await CloudConnection.Create(
+                               identity,
+                               connectionStatusChangedHandler,
+                               this.transportSettings,
+                               this.messageConverterProvider,
+                               this.clientProvider,
+                               cloudListener,
+                               this.edgeHubTokenProvider,
+                               this.idleTimeout,
+                               this.closeOnIdleTimeout,
+                               this.operationTimeout,
+                               productInfo,
+                               modelId);
+                Events.SuccessCreatingCloudConnection(identity);
+                return Try.Success(cc);
+            }
+            catch (UnauthorizedException ex) when (this.trackDeviceState)
+            {
+                return await this.TryRecoverCloudConnection(identity, connectionStatusChangedHandler, refreshOutOfDateCache, ex);
+            }
+        }
+
+        async Task<Try<ICloudConnection>> TryRecoverCloudConnection(IIdentity identity, Action<string, CloudConnectionStatus> connectionStatusChangedHandler, bool wasRefreshed, Exception ex)
+        {
+            try
+            {
+                Events.ErrorCreatingCloudConnection(identity, ex);
+                if (this.scopeAuthenticationOnly && !this.edgeHubIdentity.Id.Equals(identity.Id))
+                {
+                    if (this.trackDeviceState)
+                    {
+                        if (wasRefreshed)
+                        {
+                            Events.ErrorCreatingCloudConnection(identity, ex);
+                            return Try<ICloudConnection>.Failure(ex);
+                        }
+                        else
+                        {
+                            // recover: try to update out of date cache and try again
+                            return await this.ConnectInternalWithDeviceStateTracking(identity, connectionStatusChangedHandler, true);
+                        }
+                    }
+                    else
+                    {
+                        // old behavior
+                        Events.ErrorCreatingCloudConnection(identity, ex);
+                        return Try<ICloudConnection>.Failure(ex);
+                    }
+                }
+                else
+                {
+                    // try with cached device credentials if auth mode is not Scope or identity is for edgeHub
```

Comment:
When will this code get called? Walking back, it seems like this shoudl be called only if trackingDeviceState is true?

Reply:
it is called if it is not scope only and scope auth failed or if it is edgeHub identity and scope auth failed. trackDeviceState is always false if auth mode is not scope